soft4pes/control/lin/state_space_curr_ctr.py

```diff
@@ -0,0 +1,254 @@
+        self.save_data(ig_ref, u_k, t)
+
+        # Ensure modulating signal within -1 and 1
+        return np.clip(u_k, -1, 1)
+
+    def get_state_space_ctr_pars(self):
+        """
+        Calculate state-space controller parameters.
+        
+        Returns
+        -------
+        SimpleNamespace
+            Controller parameters.
+        """
+        Ts_pu = self.Ts_pu
+        Rf = self.Rf
+        Xf = self.Xf
+
+        # Closed-loop controller bandwidth (10x crossover frequency)
+        alpha_c = 2 * np.pi / 10 / Ts_pu
+
+        # Consider delta equals to one due to not considering delay
+        delta = 1
+
+        # Coefficients
+        phi = np.exp((-Rf / Xf) * Ts_pu) * delta
+        landa = (delta - phi) / Rf
+
+        # The closed-loop pole locations
+        p_1 = 0
+        p_2 = np.exp(-alpha_c * Ts_pu)
+        p_3 = p_2
+
+        # Coefficients
+        k_2 = -p_1 - p_2 - p_3 + phi + 1
+        k_1 = (p_1 * p_2 + p_1 * p_3 + p_2 * p_3 + k_2 * phi + k_2 -
+               phi) / landa
+
+        # State feedback gain
+        K_i = np.array([k_1, 0, k_2])
+
+        # Integral gain
+        k_ii = (-p_1 * p_2 * p_3 + k_1 * landa - k_2 * phi) / landa
+
+        # Feedforward gain
+        k_ti = k_ii / (1 - p_3)
+
+        return SimpleNamespace(delta=delta, K_i=K_i, k_ii=k_ii, k_ti=k_ti)
+
+    def state_space_controller(self, ic_dq, ic_ref_dq, uf_dq, u_max):
+        """
+        State-space controller in dq frame.
+        
+        Parameters
+        ----------
+        i_dq : 1 x 2 ndarray of floats
+            Grid Current in dq frame [p.u.].
+
+        i_ref_dq : 1 x 2 ndarray of floats
+            Reference current in dq frame [p.u.].
+
+        uf_dq : 1 x 2 ndarray of floats
+            Grid voltage in dq frame [p.u.] (In case: Without considering the filter). 
+
+        u_max : float
+            Maximum voltage.        
+
+        Returns
+        -------
+        1 x 2 ndarray of floats
+            Converter voltage reference in dq frame [p.u.].
+        """
+        #In this controller uc_km1 is not considered due to not applying delay of PWM in the state-space model
+        uc_km1 = self.uc_km1 * 0
+
+        X_LC = np.array([ic_dq, uf_dq, uc_km1])
+
+        # State space controller with anti-windup in dq frame
+        uc_ref_dq_unsat = (self.ctr_pars.k_ti * ic_ref_dq) - np.dot(
+            self.ctr_pars.K_i, X_LC) + self.u_ii_dq
+
+        u_ii_dq_ = self.ctr_pars.k_ii * ((
+            (self.uc_ref_dq - self.uc_ref_dq_unsat) / self.ctr_pars.k_ti) +
+                                         (ic_ref_dq - ic_dq))
+
+        self.u_ii_dq += u_ii_dq_
+
+        # Check for saturation
+        uc_ref_dq = self.u_saturation_check(u_max, uc_ref_dq_unsat)
+        self.uc_ref_dq = uc_ref_dq
+        self.uc_ref_dq_unsat = uc_ref_dq_unsat
+        self.uc_km1 = self.ctr_pars.delta * uc_ref_dq
+
+        return uc_ref_dq
+
+    def u_saturation_check(self, u_max, uc_ref_dq_unsat):
```

Comment:
Limiting could be used as terminology instead saturation (everywhere in the file), as that is the terminology the reference paper uses